File: lib/repositories/student.dart
import 'dart:async';
import 'dart:convert';
import 'package:coaching_management_system/repositories/globals.dart';
import 'package:http/http.dart' as http;

class Student {
  static const root = 'https://noendacademy.000webhostapp.com/student.php';
  static const _getOne = 'GET_ONE';
  static const _getAll = 'GET_ALL';
  static const _add = '_ADD';
  static const _del = '_DELETE';

  String? loginID;
  String? firstName;
  String? lastName;
  String? gender;
  String? dob;
  String? cnic;
  String? email;
  String? mobile;
  String? address;

  Student(
      {this.loginID,
      this.firstName,
      this.lastName,
      this.gender,
      this.dob,
      this.cnic,
      this.email,
      this.mobile,
      this.address});

  factory Student.fromJson(Map<String, dynamic> json) {
    return Student(
      loginID: json['loginID'] as String,
      firstName: json['First_name'] as String,
      lastName: json['Last_name'] as String,
      gender: json['Gender'] as String,
      dob: json['DOB'] as String,
      cnic: json['CNIC'] as String,
      email: json['Email'] as String,
      mobile: json['Mobile'] as String,
      address: json['Address'] as String,
    );
  }
}

Future<Student> getOneStudent(String loginID) async {
  final response = await http.post(
    Uri.parse(Student.root),
    headers: {
      'Authorization': 'Basic ${base64Encode(utf8.encode('root1:password'))}',
    },
    body: {
      'loginID': loginID,
      'action': Student._getOne,
    },
  );
  // print(response.statusCode);
  // print(response.body);
  if (jsonDecode(response.body) != "no user found") {
    return Student.fromJson(jsonDecode(response.body));
  } else {
    // If the server did not return a 201 CREATED response,
    // then throw an exception.
    //print("im running");

    throw Exception('Failed to get user.');
  }
}

Future<List<Student>> getAllStudent() async {
  try {
    final response = await http.post(
      Uri.parse(Student.root),
      headers: {
        'Authorization': 'Basic ${base64Encode(utf8.encode('root1:password'))}',
      },
      body: {
        'action': Student._getAll,
      },
    );
    //print("getEmployees >> Response:: ${response.body}");
    if (response.statusCode == 200) {
      List<Student> list = parsePhotos(response.body);
      //print(list[0].loginID);
      return list;
    } else {
      print('none found');
      throw <Student>[];
    }
  } catch (e) {
    print(e);
    return <Student>[];
  }
}

List<Student> parsePhotos(String responseBody) {
  final parsed = json.decode(responseBody).cast<Map<String, dynamic>>();
  return parsed.map<Student>((json) => Student.fromJson(json)).toList();
}

Future<String> addStudent(
    String loginID,
    String firstName,
    String lastName,
    String gender,
    String dob,
    String cnic,
    String email,
    String mobile,
    String address) async {
  try {
    final response = await http.post(
      Uri.parse(Student.root),
      headers: {
        'Authorization': 'Basic ${base64Encode(utf8.encode('root1:password'))}',
      },
      body: {
        'action': Student._add,
        'loginID': loginID,
        'first_name': firstName,
        'last_name': lastName,
        'gender': gender,
        'dob': dob,
        'cnic': cnic,
        'email': email,
        'mobile': mobile,
        'address': address,
      },
    );
    print("addEmployee >> Response:: ${response.body}");
    print(firstName);
    print(dob);
    return response.body;
  } catch (e) {
    return 'error';
  }
}

Future<String> delStudent(String loginID) async {
  try{
    final response = await http.post(
      Uri.parse(Student.root),
      headers: {
        'Authorization': 'Basic ${base64Encode(utf8.encode('root1:password'))}',
      },
      body: {
        'action': Student._del,
        'loginID': loginID,
      },
    );
    print("addEmployee >> Response:: ${response.body}");
    return response.body;
  }
  catch (e) {
    return 'error';
  }

}

File: noEndDB/student.php

<?php
    $servername = "localhost";
    $username = "root";
    $password = "";
    $dbname = "noendacademy";
    $table = "student";

    header("Access-Control-Allow-Origin: *");
    header("Access-Control-Allow-Credentials: true");
    header("Access-Control-Allow-Headers: Content-Type, Access-Control-Allow-Headers, Authorization, X-Requested-With, Access-Control-Allow-Origin");
    header("Content-Type: application/json; charset=UTF-8");
    header("Access-Control-Allow-Methods: POST, OPTIONS");



    // Create connection
    $conn = new mysqli($servername, $username, $password, $dbname);
    // Check connection
    if ($conn->connect_error) {
        die("Connection failed: " . $conn->connect_error);
        return;
    }



    $action = $_POST['action'];

    if('GET_ALL' == $action){
        $dbdata = array();
        $sql = "SELECT * FROM $table ORDER BY loginID";
        $result = $conn->query($sql);
        if ($result->num_rows > 0) {
            while($row = $result->fetch_assoc()) {
                $dbdata[]=$row;
            }
            echo json_encode($dbdata);
        } else {
            echo json_encode("no entries available");
        }
        $conn->close();
        return;
    }

    if('GET_ONE' == $action){
        $loginID = $_POST['loginID'];
        //$dbdata = array();
        $sql = "SELECT * FROM $table where loginID = '$loginID'";
        $result = $conn->query($sql);
        if ($result->num_rows > 0) {
            $row = $result->fetch_assoc();            
            echo json_encode($row);

        } else {
            echo json_encode("no user found");
        }
        $conn->close();
        return;
    }

    if('_ADD' == $action){
        $loginID = $_POST['loginID'];
        $first_name = $_POST['first_name'];
        $last_name = $_POST['last_name'];
        $gender = $_POST['gender'];
        $dob = $_POST['dob'];
        $cnic = $_POST['cnic'];
        $email = $_POST['email'];
        $mobile = $_POST['mobile'];
        $address = $_POST['address'];

        
        $query = "SELECT * from $table where loginID = '$loginID'";
        $check =  $conn->query($query);
        $count = mysqli_num_rows($check);

        if ($count == 1) {
            $msg = "user already exixts";
            //echo json_encode("User already exists!");
            echo json_encode($msg);
            $conn->close();
            return;
        }

        $sql = "INSERT INTO $table (loginID, first_name, last_name, gender, dob, cnic, email, mobile, address) 
                VALUES('$loginID', '$first_name', '$last_name', '$gender', '$dob', '$cnic', '$email', '$mobile', '$address')";
        $result = $conn->query($sql);
        echo 'success';
        $conn->close();
        return;
    }

    // if('_UPDATE' == $action){
    //     $emp_id = $_POST['emp_id'];
    //     $first_name = $_POST['first_name'];
    //     $last_name = $_POST['last_name'];
    //     $sql = "UPDATE $table SET first_name = '$first_name', last_name = '$last_name' WHERE id = $emp_id";
    //     if ($conn->query($sql) === TRUE) {
    //         echo "success";
    //     } else {
    //         echo "error";
    //     }
    //     $conn->close();
    //     return;
    // }

    if('_DELETE' == $action){
        $loginID = $_POST['loginID'];
        $sql = "DELETE FROM $table WHERE loginID = '$loginID'";
        if ($conn->query($sql) === TRUE) {
            echo json_encode("success");
        } else {
            echo json_encode("query failed");
        }
        $conn->close();
        return;
    }
    
?>
